File: scratch/check_tags.py
import re

with open("c:/Users/amotc/Documents/GitHub/repos/ChildernforLife.com/src/App.jsx", "r", encoding="utf-8") as f:
    lines = f.readlines()

# Let's extract lines 78 to 260
nav_lines = lines[77:255]

tags = []
for idx, line in enumerate(nav_lines, start=78):
    # simple regex to find tag openings and closings
    # ignoring self closing tags like <img ... /> or <path ... /> or <circle ... /> or <line ... /> or <Heart ... />
    # and ignoring JSX expressions
    # Let's find matches for:
    # 1) </tagname>
    # 2) <tagname ...> (excluding self-closing)
    
    # We will clean the line of comments
    line = re.sub(r'{\s*/\*.*?\*/\s*}', '', line)
    line = re.sub(r'//.*', '', line)
    
    # Find all tag tokens
    tokens = re.findall(r'</?([a-zA-Z0-9:-]+)|/>', line)
    for token in tokens:
        if not token:
            continue
        # Check if it is a closing tag
        is_closing = False
        match_idx = line.find(token)
        if match_idx > 0 and line[match_idx - 1] == '/':
            is_closing = True
        
        # If it is self-closing like circle, path, line, Heart, svg, etc.
        # we can ignore standard self-closing tags
        if token.lower() in ['path', 'circle', 'line', 'heart', 'img', 'br', 'hr', 'input']:
            continue
            
        if is_closing:
            if tags and tags[-1][0] == token:
                tags.pop()
            else:
                print(f"Mismatch: Closing </{token}> at line {idx} does not match open tags: {[t[0] for t in tags]}")
                tags.append((token, idx, "closing"))
        else:
            # check if it's self-closing on the same line
            # e.g. <Heart ... />
            # we'll assume it's opening for now
            tags.append((token, idx, "opening"))

print("Remaining open tags:")
for tag in tags:
    print(f"<{tag[0]}> opened at line {tag[1]}")

File: scratch/check_tags.js
const fs = require('fs');

const content = fs.readFileSync('c:/Users/amotc/Documents/GitHub/repos/ChildernforLife.com/src/App.jsx', 'utf8');
const lines = content.split('\n');

// Analyze the Navigation component block (lines 78 to 255)
const navLines = lines.slice(77, 255);

let openTags = [];
const selfClosing = new Set(['path', 'circle', 'line', 'heart', 'img', 'br', 'hr', 'input', 'meta', 'link']);

navLines.forEach((line, idx) => {
  const lineNum = idx + 78;
  
  // Clean comments
  let cleanLine = line.replace(/\{\s*\/\*.*?\*\/\s*\}/g, '').replace(/\/\/.*/g, '');
  
  // Regex to find tags: <tagname ...> or </tagname>
  // We need to match <div, </div>, etc.
  const regex = /<\/?([a-zA-Z0-9:-]+)(?:\s+[^>]*?)?(\/?)>/g;
  let match;
  
  while ((match = regex.exec(cleanLine)) !== null) {
    const tagName = match[1];
    const isClosing = cleanLine.startsWith('</' + tagName, match.index) || match[2] === '/';
    
    if (selfClosing.has(tagName.toLowerCase())) {
      continue;
    }
    
    if (isClosing) {
      if (openTags.length > 0 && openTags[openTags.length - 1].name === tagName) {
        openTags.pop();
      } else {
        console.log(`Mismatch at line ${lineNum}: closing </${tagName}> but current open tags are: ${openTags.map(t => `${t.name} (L${t.line})`).join(', ')}`);
        openTags.push({ name: tagName, line: lineNum, type: 'close-mismatch' });
      }
    } else {
      openTags.push({ name: tagName, line: lineNum, type: 'open' });
    }
  }
});

console.log("Final unmatched tags:");
openTags.forEach(t => {
  console.log(`${t.name} opened at line ${t.line}`);
});
